src/main/java/ru/yandex/practicum/filmorate/storage/db/DataBaseFilmStorage.java

```diff
@@ -0,0 +1,240 @@
+        return !filmsById.isEmpty() || !filmsByFields.isEmpty();
+    }
+
+    @Override
+    public boolean checkFilmIsPresentInStorage(Long filmId) {
+        List<Film> filmsById = filmTemplate.query(SQL_FILM_BY_ID, (rs, rowNum) -> mapFilm(rs), filmId);
+        return !filmsById.isEmpty();
+    }
+
+    @Override
+    public Long addFilmToStorage(Long filmId, Film film) {
+        Long filmIdAdded;
+        String sqlWoId = "insert into films (name, description, release_date, duration, rating_id)\n" +
+                "values(?, ?, ?, ?,\n" +
+                "(select id from ratings where rating = ?))";
+        String sqlWithId = "insert into films (id, name, description, release_date, duration, rating_id)\n" +
+                "values(?, ?, ?, ?, ?,\n" +
+                "(select id from ratings where rating = ?))";
+        if (Objects.isNull(filmId)) {
+            filmTemplate.update(sqlWoId, film.getName(), film.getDescription(),
+                    film.getReleaseDate(), film.getDuration(), film.getMpa().ratingName);
+            filmIdAdded = getLastFilmIdFromStorage();
+        } else {
+            filmTemplate.update(sqlWithId, filmId, film.getName(), film.getDescription(), film.getReleaseDate(),
+                    film.getDuration(), film.getMpa().ratingName);
+            filmIdAdded = filmId;
+        }
+        setGenresForFilmInStorage(filmIdAdded, film.getGenres());
+        return filmIdAdded;
+    }
+
+    @Override
+    public int updateFilmInStorage(Film film) {
+        String sql = "update films set name = ?, " +
+                "description = ?, " +
+                "release_date = ?, " +
+                "duration = ?, " +
+                "rating_id = ? " +
+                "where id = ?";
+        setGenresForFilmInStorage(film.getId(), film.getGenres());
+        return filmTemplate.update(sql, film.getName(), film.getDescription(),
+                film.getReleaseDate(), film.getDuration(), film.getMpa().ratingId, film.getId());
+    }
+
+    @Override
+    public boolean deleteFilmFromStorage(Long filmId, Film film) {
+        String sql = "delete from films where id = ? and name = ? and description = ?\n" +
+                "and release_date = ? and duration = ?";
+        deleteFilmGenresFromStorage(filmId);
+        return filmTemplate.update(sql, filmId, film.getName(), film.getDescription(),
+                film.getReleaseDate(), film.getDuration()) > 0;
+    }
+
+    public int deleteFilmFromStorage(Long filmId) {
+        String sql = "delete from films where id = ?";
+        deleteFilmGenresFromStorage(filmId);
+        return filmTemplate.update(sql, filmId);
+    }
+
+    @Override
+    public Long getLastFilmIdFromStorage() {
+        String sql = "select id from films order by id desc limit 1";
+        Long lastFilmId;
+        try {
+            lastFilmId = filmTemplate.queryForObject(sql, Long.class);
+        } catch (EmptyResultDataAccessException e) {
+            lastFilmId = 0L;
```

Comment:
блок catch нужен для обработки исключений. Зачем отсюда возвращать какие то значения?

Reply:
Не могу придумать как лучше обрабатывать ситуацию, когда в таблице нет записей если не через try catch?
В catch только присваивание значения в случае если при запросе из БД приходит исключение. Возвращается значение lastFilmId уже вне этого блока ниже.

Reply:
Returns:
the result object of the required type, or null in case of SQL NULL
Throws:
[IncorrectResultSizeDataAccessException](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/dao/IncorrectResultSizeDataAccessException.html) - if the query does not return exactly one row
[IncorrectResultSetColumnCountException](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/IncorrectResultSetColumnCountException.html) - if the query does not return a row containing a single column
[DataAccessException](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/dao/DataAccessException.html) - if there is any problem executing the query

Вот в документации к этому методу написано что если ничего не найдется то вернется null. DataAccessException будет если какие то проблемы с выполнением запроса. Тоесть коннекшен например отвалится. Значение будет но ты при этом ноль вернешь. Тут можно Optional использовать а если Eception то нужно его тогда обработать.